ksql-engine/src/test/java/io/confluent/ksql/analyzer/AggregateAnalyzerTest.java

```diff
@@ -15,205 +15,277 @@
+    // When:
+    analyzer.processSelect(FUNCTION_CALL);
+
+    // Then:
+    assertThat(analysis.getRequiredColumns(), contains(COL0));
+  }
+
+  @Test
+  public void shouldCaptureHavingNonAggregateFunctionArgumentsAsRequired() {
+    // When:
+    analyzer.processHaving(FUNCTION_CALL);
+
+    // Then:
+    assertThat(analysis.getRequiredColumns(), contains(COL0));
+  }
+
+  @Test
+  public void shouldCaptureGroupByNonAggregateFunctionArgumentsAsRequired() {
+    // When:
+    analyzer.processGroupBy(FUNCTION_CALL);
+
+    // Then:
+    assertThat(analysis.getRequiredColumns(), contains(COL0));
+  }
+
+  @Test
+  public void shouldCaptureSelectAggregateFunctionArgumentsAsRequired() {
+    // When:
+    analyzer.processSelect(AGG_FUNCTION_CALL);
+
+    // Then:
+    assertThat(analysis.getRequiredColumns(), contains(COL0, COL1));
+  }
+
+  @Test
+  public void shouldCaptureHavingAggregateFunctionArgumentsAsRequired() {
+    // When:
+    analyzer.processHaving(AGG_FUNCTION_CALL);
+
+    // Then:
+    assertThat(analysis.getRequiredColumns(), contains(COL0, COL1));
+  }
+
+  @Test
+  public void shouldNotCaptureNonAggregateFunction() {
+    // When:
+    analyzer.processSelect(FUNCTION_CALL);
+    analyzer.processHaving(FUNCTION_CALL);
+    analyzer.processGroupBy(FUNCTION_CALL);
+
+    // Then:
+    assertThat(analysis.getAggregateFunctions(), is(empty()));
+  }
+
+  @Test
+  public void shouldCaptureSelectAggregateFunction() {
+    // When:
+    analyzer.processSelect(AGG_FUNCTION_CALL);
+
+    // Then:
+    assertThat(analysis.getAggregateFunctions(), contains(AGG_FUNCTION_CALL));
+  }
+
+  @Test
+  public void shouldCaptureHavingAggregateFunction() {
+    // When:
+    analyzer.processHaving(AGG_FUNCTION_CALL);
+
+    // Then:
+    assertThat(analysis.getAggregateFunctions(), contains(AGG_FUNCTION_CALL));
+  }
+
+  @Test
+  public void shouldThrowOnNestedSelectAggFunctions() {
+    // Given:
+    final FunctionCall nestedCall = new FunctionCall(QualifiedName.of("MIN"),
+        ImmutableList.of(AGG_FUNCTION_CALL, COL2));
+
+    // Then:
+    expectedException.expect(KsqlException.class);
+    expectedException.expectMessage("Aggregate functions can not be nested: MIN(MAX())");
+
+    // When:
+    analyzer.processSelect(nestedCall);
+  }
+
+  @Test
+  public void shouldThrowOnNestedHavingAggFunctions() {
+    // Given:
+    final FunctionCall nestedCall = new FunctionCall(QualifiedName.of("MIN"),
+        ImmutableList.of(AGG_FUNCTION_CALL, COL2));
+
+    // Then:
+    expectedException.expect(KsqlException.class);
+    expectedException.expectMessage("Aggregate functions can not be nested: MIN(MAX())");
+
+    // When:
+    analyzer.processHaving(nestedCall);
+  }
+
+  @Test
+  public void shouldCaptureNonAggregateFunctionArgumentsWithNestedAggFunction() {
+    // Given:
+    final FunctionCall nonAggWithNestedAggFunc = new FunctionCall(QualifiedName.of("SUBSTRING"),
+        ImmutableList.of(COL2, FUNCTION_CALL, COL1));
```

Comment:
This should be AGG_FUNCTION_CALL, which breaks the test.

Reply:
Yep - already fixed as part of other work, but good catch none the less!